cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,612 @@
+                    'with reduced shape.')
+
+        # Make submodule.
+        name = 'reduce{}'.format(len(self.op_list))
+        subm = _fusion_device_func._SubmoduleReduction(
+            name, reduce_func, reduce_op, [in_param], [out_param])
+        self.submodules[subm.name] = subm
+
+        # Register Op.
+        op = _fusion_op._FusionReductionOp(subm, in_param, out_param, axes)
+        self.op_list.append(op)
+
+        # Returns.
+        return self._to_interface(out_param)
+
+    def call_indexing(self, in_param, index):
+        """Call indexing routines.
+        """
+        in_param = self._from_arraylike_interface(in_param)
+        assert isinstance(in_param, _FusionCudaArray)
+        if isinstance(index, int):
+            out_param = self.vc.index_with_int(in_param, index)
+        elif isinstance(index, tuple):
+            out_param = self.vc.index_with_tuple(in_param, index)
+        else:
+            raise NotImplementedError
+
+        return self._to_interface(out_param)
+
+    def _call_function(self, func, args):
+        # Registers input variables.
+        in_params = []
+        for input_order, arg in enumerate(args):
+            if arg is None:
+                var = None
+            elif isinstance(arg, core.ndarray):
+                same_array_indices = [
+                    i for i in range(input_order) if arg is args[i]]
+                may_share_memory_indices = [
+                    i for i in range(input_order)
+                    if (isinstance(args[i], core.ndarray)
+                        and _base(arg) is _base(args[i]))]
+
+                if len(same_array_indices) > 0:
+                    i = same_array_indices[0]
+                    assert isinstance(in_params[i], _FusionCudaArray)
+                    var = in_params[i]
+                elif len(may_share_memory_indices) > 0:
+                    i = may_share_memory_indices[0]
+                    assert isinstance(in_params[i], _FusionCudaArray)
+                    var = self.vc.make_view(
+                        in_params[i], input_order=input_order)
+                else:
+                    var = self.vc.generate_new_array(
+                        arg.dtype, arg.shape, None, input_order=input_order)
+            else:
+                dtype = numpy.dtype(type(arg))
+                var = self.vc.generate_new_scalar(
+                    dtype, input_order=input_order)
+            in_params.append(var)
+
+        # Calls the target function.
+        out_params = func(*[self._to_interface(x) for x in in_params])
+
+        # Registers output variables.
+        if isinstance(out_params, tuple):
+            return_size = len(out_params)
+            out_params = [self._from_interface(x) for x in out_params]
+        elif out_params is None:
+            return_size = 'none'
+            out_params = []
+        else:
+            return_size = 'single'
```

Comment:
Do you have to distinguish between `'none'` and `'single'`?
How about using just `'single'`?